### DISCUSSION TITLE
How do I make photons run multiple tile animations in a loop?

### DISCUSSION THREAD

**Discussion by danktankk** [delfick/photons#106] [Q&A] [answered]
After reading the notes several times, I will definitely need some help lol.  I know you said documentation will probably never be done for this for animations, but better documentation would lead to more people being able to use it I would think..  Your call though :)

I am mainly just looking for a way to get the same colors as energizing for tiles in the LiFX app working for both `falling` and `twinkles` with a way to dim it as well, then add `swipe` after x seconds, then `twinkles` with `swipe` again and back to `falling` and loop that.  Any ideas?

**Comment by delfick** [comment, marked as answer]
> but better documentation would lead to more people being able to use it I would think

You're correct and I agree. That's not the problem though

>  Your call though :)

I'm considerably burnt out by LIFX stuff and also my mental health in general is currently ruined and not overly coping with a lot of things. So whilst I'm more than happy to answer questions and do small fixes I don't have the capacity for any large projects on this one anytime soon.

> Any ideas?

So when I rewrote the animations I very specifically made it so that you can tell Photons to run multiple animations and have them transition between them :)

For example if you save this json file  https://gist.github.com/delfick/286e7518d98f23a5a2503ff6f9a7f3a7

and run

```
lifx lan:animate -- file://instructions.json
```

And you end up filling in this object https://github.com/delfick/photons/blob/main/modules/photons_canvas/animations/run_options.py#L112

So that json file says keeping going through the list of animations until we've had 6 animations (so if it only had 2 items in the list, it'd repeat those 2 three times)

And then the animations in the json are:

* color_cycle animation using its "vertical_morph" changer for 7 seconds starting from whatever pixels were already set on the tiles
* same but the wave changer
* etc
* swipe animation where photons thinks the tiles are rotated in ways they currently aren't until it's done

You'll notice each item in that list is an array with a name (the name of the animation) and a dictionary of options. The options are defined by each animation so they all differ a bit. But there are a few common options that apply to everything - https://github.com/delfick/photons/blob/e5898f1/modules/photons_canvas/animations/infrastructure/animation.py#L26-L33

* every - float (default 0.075, but is overridden by some animations) - seconds between each update to the tiles (an animate is sending 64 pixels to each panel every `every` seconds)
* retries - bool (default False) - whether to care about whether messages are acknowledged by the tiles (slower but can be important to some animations)
* duration - float (default 0) - Seconds the firmware should take to fade when we give it new pixels
* num_seconds - Nullable float (default None) - whether there is a time limit in seconds for the duration of the animation (some animations loop forever, some will naturally finish, like swipe)
* message_timeout - float (default 0.3) - seconds we wait when retries are on and a message isn't ack'd
* random_orientations - bool (default False) - whether to make photons pretend you've rotated your tiles randomly
* skip_next_transition - bool (default False) the run options let you say there should be a transition animation between everything, this option lets you skip that for a particular animation.

Beyond those options you can work out options each animation is by saying

```
lifx lan:animate help color_cycle
```

Where you replace color_cycle by the name of the animation

```
lifx lan:animate help swipe
```

to find all the available animations in the first place run

```
lifx lan:animate help
```

I'm not at home today and getting out the tiles I have is a pain anyway, but a good starting point for you to start just throwing words at your computer would be something like

```
{
  "animations": [
    [
      "falling",
      {
        "line_hues": "0-360,0.2-0.5",
        "num_seconds": 7
      }
    ],
    [
      "twinkles",
      {
        "twinkles_color_range": "0-360,0.2-0.5",
        "num_seconds": 7
      }
    ],
    ["swipe"],
    [
      "twinkles",
      {
        "twinkles_color_range": "0-360,0.2-0.5",
        "num_seconds": 7
      }
    ],
    ["swipe"],
    [
      "falling",
      {
        "line_hues": "0-360,0.2-0.5",
        "num_seconds": 7
      }
    ],
    ["swipe"]
  ]
}
```

**Comment by danktankk** [comment]
Thanks a ton for the help and links!  I have the pattern set how I would like by using the json file you linked and then fiddled with it until I got it how I like it.

I cant figure out what the exact colors are for the `energizing theme` or how I would add them specifically to the json file.

Thanks again for getting me this far!

**Reply by delfick** [reply]
awesome :)

As for the theme, there's no public way of getting the hue values in a theme. The LIFX support might be able to tell you though if you [contact](https://support.lifx.com/) them.

Then you'd specify with something like `"line_hues": ["h,s,b,k", "h,s,b,k", ...]`

Actually, just realised that means my example above was setting saturation, not brightness. Where I say `0-360,0.2-0.5` I mean `0-360,1,0.2-0.5`. I forget saturation comes first.

So if the theme is `[(50, 1, 0.5, 3500), (60, 1, 0.5, 3500), (70, 1, 0.5, 3500)]` then you'd want `"line_hues": ["50,1,0.5", "60,1,0.5", "70,1,0.5"]` (photons already defaults the kelvin to 3500)

The dash lets you make a range. So when deciding a color to use, photons will look at that list, randomly choose one of them, and then randomly choose a value from the range. Which is why `0-360` results in any hue value. https://lan.developer.lifx.com/docs/representing-color-with-hsbk

**Reply by danktankk** [reply]
thanks for the reply.  I was hoping there was a way to sniff the packet data to see the values.  bummer....

**Reply by Djelibeybi** [reply]
I know this is a long time ago, but https://github.com/wborn/lifx-themes has a good breakdown of the colours used for each of the original themes. 

You can get the colour values locally by setting a theme on your Tiles using the app and then introspecting the state of the Tiles using `lifx lan:get_chain_state` which will output the colour of each pixel on each Tile.
